adapter/internal/discovery/xds/server.go

```diff
@@ -466,6 +520,109 @@ func UpdateAPI(vHost string, apiProject mgw.ProjectAPI, deployedEnvironments []*
+		if (isMajorRangeRegexAvailable || isMinorRangeRegexAvailable) && (isLatestMajorVersion || isLatestMinorVersion) {
+			// Organization's all apis
+			for apiUUID, swagger := range orgIDAPIMgwSwaggerMap[organizationID] {
+				// API's all versions in the same vHost
+				if swagger.GetTitle() == apiName && swagger.GetVHost() == vHost {
+					if isMajorRangeRegexAvailable && swagger.GetVersion() == existingMajorRangeLatestVersion ||
+						isMinorRangeRegexAvailable && swagger.GetVersion() == existingMinorRangeLatestVersion {
+						for _, route := range orgIDOpenAPIRoutesMap[organizationID][apiUUID] {
+							regex := route.GetMatch().GetSafeRegex().GetRegex()
+							regexRewritePattern := route.GetRoute().GetRegexRewrite().GetPattern().GetRegex()
+							if isMinorRangeRegexAvailable && swagger.GetVersion() == existingMinorRangeLatestVersion && isLatestMinorVersion {
+								regex = strings.ReplaceAll(regex, GetMinorVersionRangeRegex(existingMinorRangeLatestVersion), existingMinorRangeLatestVersion)
+								regex = strings.ReplaceAll(regex, GetMajorMinorVersionRangeRegex(existingMajorRangeLatestVersion), existingMajorRangeLatestVersion)
+								regexRewritePattern = strings.ReplaceAll(regexRewritePattern, GetMinorVersionRangeRegex(existingMinorRangeLatestVersion), existingMinorRangeLatestVersion)
+								regexRewritePattern = strings.ReplaceAll(regexRewritePattern, GetMajorMinorVersionRangeRegex(existingMajorRangeLatestVersion), existingMajorRangeLatestVersion)
+							}
+							if isMajorRangeRegexAvailable && swagger.GetVersion() == existingMajorRangeLatestVersion && isLatestMajorVersion {
+								regex = strings.ReplaceAll(regex, GetMajorMinorVersionRangeRegex(existingMajorRangeLatestVersion), GetMinorVersionRangeRegex(existingMajorRangeLatestVersion))
+								regexRewritePattern = strings.ReplaceAll(regexRewritePattern, GetMajorMinorVersionRangeRegex(existingMajorRangeLatestVersion), GetMinorVersionRangeRegex(existingMajorRangeLatestVersion))
+							}
+							pathSpecifier := &routev3.RouteMatch_SafeRegex{
+								SafeRegex: &envoy_type_matcherv3.RegexMatcher{
+									Regex: regex,
+								},
+							}
+							route.Match.PathSpecifier = pathSpecifier
+							action := &routev3.Route_Route{}
+							action = route.Action.(*routev3.Route_Route)
+							action.Route.RegexRewrite.Pattern.Regex = regexRewritePattern
+							route.Action = action
+						}
+					}
+				}
+			}
+		}
+
+		if isLatestMajorVersion || isLatestMinorVersion {
+			// Update local memory map with the latest version ranges
+			if _, ok := orgIDLatestAPIVersionMap[organizationID]; !ok {
+				latestVersions := make(map[string]string)
+				latestVersions[GetMajorVersionRange(apiVersion)] = apiVersion
+				latestVersions[GetMinorVersionRange(apiVersion)] = apiVersion
+				apiVersionMap := make(map[string]map[string]string)
+				apiVersionMap[apiRangeIdentifier] = latestVersions
+				orgIDLatestAPIVersionMap[organizationID] = apiVersionMap
```

Comment:
When APIs are getting undeployed, doesn't the entries in this map is removed?

Reply:
Eg:

When `1.2.0` is undeployed.
`/foo/v1` ---> 1.2.0 should be changed to `/foo/v1` ---> 1.0.2

Reply:
Added routing updating on API undeployment with: https://github.com/wso2/product-microgateway/pull/3376/commits/a7a452c60f9eddc276ecb9713e1707cb9d9415cb